File: crates/c-api/tests/global.cc
#include <wasmtime/global.hh>

#include <gtest/gtest.h>
#include <wasmtime.hh>

using namespace wasmtime;

TEST(Global, Smoke) {
  Engine engine;
  Store store(engine);
  Global::create(store, GlobalType(ValKind::I32, true), 3.0).err();
  Global::create(store, GlobalType(ValKind::I32, true), 3).unwrap();
  Global::create(store, GlobalType(ValKind::I32, false), 3).unwrap();

  Global g = Global::create(store, GlobalType(ValKind::I32, true), 4).unwrap();
  EXPECT_EQ(g.get(store).i32(), 4);
  g.set(store, 10).unwrap();
  EXPECT_EQ(g.get(store).i32(), 10);
  g.set(store, 10.23).err();
  EXPECT_EQ(g.get(store).i32(), 10);

  EXPECT_EQ(g.type(store)->content().kind(), ValKind::I32);
  EXPECT_TRUE(g.type(store)->is_mutable());
}

File: crates/fuzzing/tests/oom/global.rs
#![cfg(arc_try_new)]

use wasmtime::{Config, Engine, Global, GlobalType, Mutability, Result, Store, Val, ValType};
use wasmtime_fuzzing::oom::OomTest;

#[test]
fn global_new() -> Result<()> {
    let mut config = Config::new();
    config.enable_compiler(false);
    config.concurrency_support(false);
    let engine = Engine::new(&config)?;

    OomTest::new().test(|| {
        let mut store = Store::try_new(&engine, ())?;
        let ty = GlobalType::new(ValType::I32, Mutability::Var);
        let _global = wasmtime::Global::new(&mut store, ty, Val::I32(42))?;
        Ok(())
    })
}

#[test]
fn global_get() -> Result<()> {
    let mut config = Config::new();
    config.enable_compiler(false);
    config.concurrency_support(false);
    let engine = Engine::new(&config)?;

    OomTest::new().test(|| {
        let mut store = Store::try_new(&engine, ())?;
        let ty = GlobalType::new(ValType::I32, Mutability::Var);
        let global = Global::new(&mut store, ty, Val::I32(42))?;
        let val = global.get(&mut store);
        assert_eq!(val.unwrap_i32(), 42);
        Ok(())
    })
}

#[test]
fn global_set() -> Result<()> {
    let mut config = Config::new();
    config.enable_compiler(false);
    config.concurrency_support(false);
    let engine = Engine::new(&config)?;

    OomTest::new().test(|| {
        let mut store = Store::try_new(&engine, ())?;
        let ty = GlobalType::new(ValType::I32, Mutability::Var);
        let global = Global::new(&mut store, ty, Val::I32(42))?;
        global.set(&mut store, Val::I32(99))?;
        let val = global.get(&mut store);
        assert_eq!(val.unwrap_i32(), 99);
        Ok(())
    })
}

#[test]
fn global_ty() -> Result<()> {
    let mut config = Config::new();
    config.enable_compiler(false);
    config.concurrency_support(false);
    let engine = Engine::new(&config)?;

    OomTest::new().test(|| {
        let mut store = Store::try_new(&engine, ())?;
        let ty = GlobalType::new(ValType::I32, Mutability::Var);
        let global = Global::new(&mut store, ty, Val::I32(42))?;
        let ty = global.ty(&store);
        assert!(ty.content().is_i32());
        Ok(())
    })
}
